eclair-core/src/test/scala/fr/acinq/eclair/io/IncomingConnectionsTrackerSpec.scala

```diff
@@ -0,0 +1,115 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package fr.acinq.eclair.io
+
+import akka.actor.testkit.typed.scaladsl.{ScalaTestWithActorTestKit, TestProbe}
+import akka.actor.typed.ActorRef
+import akka.actor.typed.eventstream.EventStream
+import akka.actor.typed.scaladsl.Behaviors
+import akka.actor.typed.scaladsl.adapter.TypedActorRefOps
+import com.typesafe.config.ConfigFactory
+import fr.acinq.bitcoin.scalacompat.Crypto
+import fr.acinq.eclair.TestConstants.Alice.nodeParams
+import fr.acinq.eclair.channel.ChannelOpened
+import fr.acinq.eclair.io.IncomingConnectionsTracker.{ForgetIncomingConnection, TrackIncomingConnection}
+import fr.acinq.eclair.io.Peer.Disconnect
+import fr.acinq.eclair.{randomBytes32, randomKey}
+import org.scalatest.Outcome
+import org.scalatest.funsuite.FixtureAnyFunSuiteLike
+
+import scala.concurrent.duration.DurationInt
+
+class IncomingConnectionsTrackerSpec extends ScalaTestWithActorTestKit(ConfigFactory.load("application")) with FixtureAnyFunSuiteLike {
+  val connection1: Crypto.PublicKey = randomKey().publicKey
+  val connection2: Crypto.PublicKey = randomKey().publicKey
+
+  override def withFixture(test: OneArgTest): Outcome = {
+    val nodeParams1 = nodeParams.copy(peerConnectionConf = nodeParams.peerConnectionConf.copy(maxNoChannels = 2))
+    val switchboard = TestProbe[Disconnect]()
+    val monitorProbe = testKit.createTestProbe[IncomingConnectionsTracker.Command]()
+    val tracker = testKit.spawn(Behaviors.monitor(monitorProbe.ref, IncomingConnectionsTracker(nodeParams1, switchboard.ref)))
+    withFixture(test.toNoArgTest(FixtureParam(tracker, switchboard, monitorProbe)))
+  }
+
+  case class FixtureParam(tracker: ActorRef[IncomingConnectionsTracker.Command], switchboard: TestProbe[Disconnect], monitorProbe: TestProbe[IncomingConnectionsTracker.Command])
+
+  test("accept new node connections, after limit is reached kill oldest node connection first") { f =>
+    import f._
+
+    tracker ! IncomingConnectionsTracker.TrackIncomingConnection(connection1)
+    tracker ! IncomingConnectionsTracker.TrackIncomingConnection(connection2)
+    tracker ! IncomingConnectionsTracker.TrackIncomingConnection(randomKey().publicKey)
+    assert(switchboard.expectMessageType[Disconnect].nodeId === connection1)
+    tracker ! IncomingConnectionsTracker.TrackIncomingConnection(randomKey().publicKey)
+    assert(switchboard.expectMessageType[Disconnect].nodeId === connection2)
+  }
+
+  test("stop tracking a node that disconnects and free space for a new node connection") { f =>
+    import f._
+
+    // Track nodes without channels.
+    tracker ! IncomingConnectionsTracker.TrackIncomingConnection(connection1)
+    tracker ! IncomingConnectionsTracker.TrackIncomingConnection(connection2)
+    monitorProbe.expectMessageType[TrackIncomingConnection]
+    monitorProbe.expectMessageType[TrackIncomingConnection]
```

Comment:
I don't understand the goal of these assertions, they aren't testing anything, are they?
If you want to make sure that the two messages above have been received, you should rather do:

```scala
    eventually {
      tracker ! IncomingConnectionsTracker.CountIncomingConnections(probe.ref)
      probe.expectMessage(2)
    }
```

Same elsewhere, the `monitorProbe` assertions seem useless...I believe you initially used it to test that the actor crashed if sent `Status.Failure`, but you can probably remove `monitorProbe` entirely?

Reply:
I forgot now why I reintroduced that; I think I was looking for a way to detect when the tracker dies due to the original issue you found about receiving an unhandled message from `switchboard`. Anyway, it doesn't need a test with the current fix because now `Peer.Disconnect` never has a reference to the `incomingConnectionsTracker`. Removed in [a00f448](https://github.com/ACINQ/eclair/pull/2601/commits/a00f448df3d45b3ab26952948a3c47f6cfeec9c0).